### PR TITLE
[java] Fix false negative problem about Enum in AvoidFieldNameMatchingMethodName #3936

### LINKED ISSUE #3936: [java] AvoidFieldNameMatchingMethodName should consider enum class
**Affects PMD Version:6.42.0** 

**Rule:**

Please provide the rule name and a link to the rule documentation:
<https://pmd.github.io/latest/pmd_rules_java_errorprone.html#avoidfieldnamematchingmethodname>

**Description:**
Hi, I found a false negative about the rule AvoidFieldNameMatchingMethodName. PMD should have reported a warning in line 5. I think this bug is caused by enum class here. Thanks for your consideration.

**Code Sample demonstrating the issue:**
```java
public class C {
  enum E {
      R;
      int bar;
      void bar() {}  // should report a warning in this line
  }
}
```

**Expected outcome:**

PMD should report a violation at line 5, but doesn't. This is a false-negative.

**Running PMD through:** *[Maven]*

### PR DESCRIPTION
**Pull request by Scrsloota**
## Describe the PR

This PR fixes a false negative problem about Enum in the rule AvoidFieldNameMatchingMethodName encountered in issue 3936. This problem has been fixed by changing the methods in AvoidFieldNameMatchingMethodNameRule.java file. All added tests are passed.

## Related issues

- Fixes #3936

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 23 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 6 configuration errors.<br />
<a href="https://chunk.io/pmd/36dd89636e1e4d5db8d1dd8833defefc/diff1/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java [hunk 1]
```diff
@@ -28,29 +31,36 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
...
-            if (child.getNumChildren() == 0) {
-                continue;
-            }
-            child = child.getChild(child.getNumChildren() - 1);
-            if (child instanceof ASTFieldDeclaration) {
-                fields.add((ASTFieldDeclaration) child);
-            } else if (child instanceof ASTMethodDeclaration) {
-                methodNames.add(((ASTMethodDeclaration) child).getName().toLowerCase(Locale.ROOT));
+    public Object visit(ASTEnumDeclaration node, Object data) {
+        return super.visit(node, data);
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java:35]: This method now merely calls super, so we don't need to override it

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java [hunk 2]
```diff
@@ -28,29 +31,36 @@ public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
...
+    }
+
+    @Override
+    public Object visit(JavaNode node, Object data) {
+        if (node instanceof ASTClassOrInterfaceBody || node instanceof ASTEnumBody) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java:40]: This is a very uncommon pattern. Usually we do this in such a case:

```java
public Object visit(ASTClassOrInterfaceBody node, Object data) {
    handleClassOrEnum(node, data);
    return super.visit(node, data);
}

public Object visit(ASTEnumBody node, Object data) {
    handleClassOrEnum(node, data);
    return super.visit(node, data);
}

private void handleClassOrEnum(JavaNode node, Object data) {
    // impl here
}
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/AvoidFieldNameMatchingMethodName.xml [hunk 3]
```diff
@@ -60,6 +60,48 @@ public class Bar {
         <code><![CDATA[
 public interface Bar {
     public static final int FOO = 5;
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Test1 in Enum</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/AvoidFieldNameMatchingMethodName.xml:68]:
```suggestion
        <description>Test1 in Enum #3936</description>
```

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for the PR!

The regression test reports looks good - all new reported violation are indeed correct.
Please change the implementations to avoid the instanceof checks.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java (hunk 1 above)
Comment (Scrsloota) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java:35]: I see. I'll remove this method. 😊

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java (hunk 2 above)
Comment (Scrsloota) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java]: Thank you for your instruction. I will rewrite this part!

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/AvoidFieldNameMatchingMethodName.xml (hunk 3 above)
Comment (Scrsloota) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/AvoidFieldNameMatchingMethodName.xml]: I will add the message of issue into these three test cases. 😃

**Review by adangel — APPROVED** [review]
Thanks, looks good!

**Pull request merged**